lib/AST/ASTImporter.cpp

```diff
@@ -7000,6 +6993,8 @@ Decl *ASTImporter::GetAlreadyImportedOrNull(Decl *FromD) {
   if (Pos != ImportedDecls.end()) {
     Decl *ToD = Pos->second;
     ASTNodeImporter(*this).ImportDefinitionIfNeeded(FromD, ToD);
+    if (FromD->isUsed(false) && !ToD->isUsed(false))
```

Comment:
Why is this change in `GetAlreadyImportedOrNull` ?
I think this change should have a better place in `void InitializeImportedDecl(Decl *FromD, Decl *ToD)`. There is already a `setIsUsed()` call there:
```
    void InitializeImportedDecl(Decl *FromD, Decl *ToD) {
      Importer.MapImported(FromD, ToD);
      ToD->IdentifierNamespace = FromD->IdentifierNamespace;
      if (FromD->hasAttrs())
        for (Attr *FromAttr : FromD->getAttrs())
          ToD->addAttr(FromAttr->clone(ToD->getASTContext()));
      if (FromD->isUsed())
        ToD->setIsUsed();   // <------------------------- Here
      if (FromD->isImplicit())
        ToD->setImplicit();
    }
```

To have only a simple responsibility and to comply to the name, `GetAlreadyImportedOrNull` should take it's parameter as a `Decl const *` and it could be a const member function.

Reply:
`isUsed(false)` skips the `used` attribute. Why do we want to skip that?

Reply:
The `used` flag is set on the canonical decl. Perhaps the crux of the problem is that we do not import the declaration chain correctly in case of `VarDecl`s? Could you please check this?

Reply:
The problem is that the isUsed flag is changed (for some reason) for the original `Decl` after it was first imported. Probably lldb uses the `ASTImporter` in an unexpected way. The `Decl` is first visited (without a `DeclRefExpr` and in non-used state), then the `DeclRefExpr` that contains the same `Decl` object and the 'used' flag is set to true here. The same `Decl` object is tried to be imported more than once and the isUsed is not true in the first import when `InitializeImportedDecl` is called. (The `GetAlreadyImportedOrNull` can be modified to not change anything in the imported object, but then the `ImportDefinitionIfNeeded` call should be moved to other place too.)

Reply:
Using the `used` attribute in `isUsed` does not influence the non-attributed the meaning, so I think we should keep that.
```
bool Decl::isUsed(bool CheckUsedAttr) const {
  const Decl *CanonD = getCanonicalDecl();
  if (CanonD->Used)
    return true;

  // Check for used attribute.
  // Ask the most recent decl, since attributes accumulate in the redecl chain.
  if (CheckUsedAttr && getMostRecentDecl()->hasAttr<UsedAttr>())
    return true;

  // The information may have not been deserialized yet. Force deserialization
  // to complete the needed information.
  return getMostRecentDecl()->getCanonicalDecl()->Used;
}
```

Reply:
Import of `VarDecl` does not create declaration chain. Probably something is not correct with this. But in the concrete case there should be no chain, there is a single file-level declaration of a variable.